File: src/ksana_llm/python/serving_forward_client.py
# Copyright 2024 Tencent Inc.  All rights reserved.
#
# ==============================================================================

import multiprocessing
import argparse
import time
import base64

import msgpack
import requests
import numpy as np


def args_config():
    parser = argparse.ArgumentParser()
    parser.add_argument('--host',
                        type=str,
                        default="localhost",
                        help='server host address')
    parser.add_argument('--port', type=int, default=8888, help='server port')
    parser.add_argument('--api', type=str, default="forward", help='server api')
    args = parser.parse_args()
    return args


def post_request_msgpack(serv, data, queue=None):
    packed_data = msgpack.packb(data)
    responses = requests.post(serv, data=packed_data, headers={'Content-Type': 'application/x-msgpack'})
    if responses.status_code == 200:
        result = msgpack.unpackb(responses.content)
    else:
        result = f"Failed to get response: {responses}"
    if queue is None:
        return result
    else:
        queue.put(result)


def python_tensor_to_numpy(python_tensor):
    """
    Converts a custom PythonTensor object into a NumPy array.

    This function is designed to transform a PythonTensor, which is a custom object for this example,
    into a NumPy array. The conversion process takes into account the specified shape and
    data type of the input tensor to ensure accurate representation in the NumPy array format.

    Parameters:
    - python_tensor: A dictionary containing the data buffer from the PythonTensor object,
                     the desired shape for the resulting NumPy array, and the desired data type.
                     The data buffer should be a buffer or a similar object that
                     holds the raw data.

    Returns:
    - numpy.ndarray: A NumPy array constructed from the PythonTensor's data, adhering to the
                     specified shape and data type requirements.

    Raises:
    - ValueError: If the specified data type is not recognized. Currently, only 'float32',
                  'float16', 'uint16', and 'int32' data types are supported, represented as
                  'float32', 'float16', 'bfloat16', and 'int32' respectively.
    """
    # Extract data, shape, and dtype from the python_tensor dictionary
    data, shape, dtype = python_tensor["data"], python_tensor["shape"], python_tensor["dtype"]

    # Map the specified string data type to its corresponding NumPy data type
    if dtype == "float32":
        np_dtype = np.float32
    elif dtype == "float16":
        np_dtype = np.float16
    elif dtype == "bfloat16":
        np_dtype = np.uint16
    elif dtype == "int32":
        np_dtype = np.int32
    else:
        # Throw an exception for unsupported data types
        raise ValueError(f"Unsupported dtype: {dtype}")

    # Create a NumPy array from the raw data buffer using the specified data type
    data_array = np.frombuffer(base64.b64decode(data), dtype=np_dtype)

    # Reshape the NumPy array according to the specified shape
    numpy_array = data_array.reshape(shape)

    # Special handling for 'bfloat16' dtype to convert uint16 to float32
    if dtype == "bfloat16":
        numpy_array = numpy_array.astype(np.uint32) << 16
        numpy_array = numpy_array.view(np.float32)

    return numpy_array


def show_response(result):
    if isinstance(result, dict) and "responses" in result:
        for batch_result in result["responses"]:
            if isinstance(batch_result, dict) and "response" in batch_result:
                input_token_ids = batch_result["input_token_ids"]
                for response in batch_result["response"]:
                    target = response["target_name"]
                    python_tensor = response["tensor"]
                    print(
                        f"input_token_ids : {input_token_ids}, target : {target}, "
                        f"tensor : \n{python_tensor_to_numpy(python_tensor)}"
                    )
    else:
        print(result)


if __name__ == "__main__":
    args = args_config()
    serv = "http://" + args.host + ":" + str(args.port) + "/" + args.api

    text_list = [
        "<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\n你好。<|im_end|>\n"
        "<|im_start|>assistant",
    ]

    multi_proc_list = []
    multi_proc_queue = multiprocessing.Queue()
    start_time = time.time()
    # Iterate over the list of text prompts
    for i in range(len(text_list)):
        # Create a prompt string from the current text element
        prompt = "%s" % text_list[i]

        # Create a data dictionary to pass to the post_request function
        data = {
            "requests": [  # Set requests, multiple are allowed
                {
                    "prompt": prompt,  # Set the prompt for the request
                    # "input_tokens": [1,22,13],  # Or set the input tokens for the request
                    # "input_refit_embedding": {
                    #     "pos": [20,285,550],
                    #     "embeddings": [[1.0,2.0,...],[3.0,4.0,...],[5.0,6.0,...]],
                    # },
                    "request_target": [  # Specify the target, multiple are allowed
                        {
                            "target_name": "layernorm",  # The target can be "layernorm", "transformer" or "logits"
                            "token_id": [
                                13
                            ],  # Provide specific token ids (only permitted in "layernorm" and "transformer")
                            # "slice_pos" : [[0,0],[2,5],[7,7]], # Set sorted intervals (negative indices are allowed)
                            "token_reduce_mode": "GATHER_ALL",  # For "layernorm" and "transformer"
                            # "token_reduce_mode" : "GATHER_TOKEN_ID", # For "logits"
                        },
                    ],
                },
            ]
        }

        # Create a new process to handle the post request
        proc = multiprocessing.Process(
            target=post_request_msgpack,  # function to call
            args=(  # arguments to pass
                serv,  # server object
                data,  # data dictionary
                multi_proc_queue,  # queue for communication
            ))
        # Start the process
        proc.start()
        # Add the process to the list of processes
        multi_proc_list.append(proc)

    # Wait for all processes to finish
    for proc in multi_proc_list:
        # Join the process to ensure it has finished
        proc.join()
    end_time = time.time()

    # Get the response of each process from the queue and display it
    for i in range(len(text_list)):
        show_response(multi_proc_queue.get())

    print("{} requests duration: {:.3f}s".format(len(text_list),
                                                 end_time - start_time))

File: src/ksana_llm/cpp/serving_forward_client.cpp
/* Copyright 2024 Tencent Inc.  All rights reserved.
==============================================================================*/

#include <getopt.h>
#include <cassert>
#include <chrono>
#include <cstring>
#include <iostream>
#include <vector>

#include "base64.hpp"
#include "httplib.h"
#include "ksana_llm/utils/request_serial.h"
#include "msgpack.hpp"

struct Args {
  // server host address
  const char *host = "localhost";
  // server port
  int port = 8888;
  // server api
  const char *api = "forward";
};

void PrintUsage(char *program) {
  std::cout << "usage: " << program << " [-h, --help] [-s, --host HOST] [-p, --port PORT] [-a, --api API]\n\n";
  std::cout << "optional arguments:\n";
  std::cout << "  -h, --help       show this help message and exit\n";
  std::cout << "  -s, --host HOST  server host address\n";
  std::cout << "  -p, --port PORT  server port\n";
  std::cout << "  -a, --api  API   server api\n";
  exit(1);
}

// Parse the command line arguments.
Args ParseArgs(int argc, char *argv[]) {
  Args args;
  const option options[] = {{"host", required_argument, nullptr, 's'},
                            {"port", required_argument, nullptr, 'p'},
                            {"api", required_argument, nullptr, 'a'},
                            {"help", no_argument, nullptr, 'h'},
                            {nullptr, 0, nullptr, 0}};

  int opt;
  while ((opt = getopt_long(argc, argv, "s:p:a:h", options, nullptr)) != -1) {
    switch (opt) {
      case 's':  // -s or --host
        args.host = optarg;
        break;
      case 'p':  // -p or --port
        args.port = std::stoi(optarg);
        break;
      case 'a':  // -a or --api
        args.api = optarg;
        break;
      case 'h':  // -h or --help
      case '?':  // unrecognized option
      default:
        PrintUsage(argv[0]);
    }
  }
  return args;
}

// Send request packed in msgpack to the server.
bool PostRequestMsgPack(httplib::Client &cli, const std::string &api, const ksana_llm::BatchRequestSerial &request,
                        ksana_llm::BatchResponseSerial &response) {
  msgpack::sbuffer sbuf;
  msgpack::pack(sbuf, request);

  if (auto rsp = cli.Post("/" + api, sbuf.data(), sbuf.size(), "application/x-msgpack"); !rsp || rsp->status != 200) {
    std::cerr << "Failed to get response: " << rsp << std::endl;
    return false;
  } else {
    auto handle = msgpack::unpack(rsp->body.data(), rsp->body.size());
    auto object = handle.get();
    object.convert(response);
    return true;
  }
}

int main(int argc, char *argv[]) {
  std::cout << std::fixed << std::setprecision(5);

  Args args = ParseArgs(argc, argv);

  // Create an HTTP client to the server.
  httplib::Client cli(args.host, args.port);

  auto start = std::chrono::high_resolution_clock::now();
  // Prepare the request data.
  ksana_llm::BatchRequestSerial request;
  request.requests.push_back(
      ksana_llm::RequestSerial{.prompt = "Hello, world!",  // Set the prompt.
                               .request_target = {ksana_llm::TargetRequestSerial{
                                   // Set a list of targets.
                                   .target_name = "logits",  // Request logits output.
                                   .slice_pos = {{0, 1}},    // Provide a set of sorted intervals of the result.
                                   .token_reduce_mode = "GATHER_TOKEN_ID"}}});
  request.requests.push_back(
      ksana_llm::RequestSerial{.input_tokens = {1, 22, 13},  // Set the input tokens.
                               .request_target = {ksana_llm::TargetRequestSerial{
                                   // Set a list of targets.
                                   .target_name = "logits",  // Request logits output.
                                   .slice_pos = {{-3, -2}},  // Indices can be negative to count from the end.
                                   .token_reduce_mode = "GATHER_TOKEN_ID"}}});

  // Send request to the server and get the response.
  ksana_llm::BatchResponseSerial response;
  PostRequestMsgPack(cli, args.api, request, response);
  auto end = std::chrono::high_resolution_clock::now();

  // Parse the response and show the result.
  for (auto &each_response : response.responses) {
    for (auto &target_response : each_response.response) {
      auto &tensor = target_response.tensor;
      // C++ does not have primitive half precision float type.
      assert(tensor.dtype == "float32");
      auto data_bytes = base64::decode_into<std::vector<uint8_t>>(tensor.data.begin(), tensor.data.end());
      assert(tensor.shape[0] * sizeof(float) == data_bytes.size());
      std::vector<float> data(tensor.shape[0]);
      memcpy(data.data(), data_bytes.data(), data_bytes.size());

      std::cout << "input_token_ids : [ ";
      for (auto token : each_response.input_token_ids) {
        std::cout << token << " ";
      }
      std::cout << "], target_name : " << target_response.target_name << ", tensor : \n";
      std::cout << "[ ";
      for (auto value : data) {
        std::cout << value << " ";
      }
      std::cout << "]" << std::endl;
    }
  }

  auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
  std::cout << "request duration: " << duration << " ms" << std::endl;

  return 0;
}
